### PR TITLE
feat(manager:terraform): Terraform-module updates for Azure DevOps SSH Repositories

### LINKED ISSUE #15491: Terraform modules with SSH not supported for Azure DevOps repositories
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

32.45.2

### Please select which platform you are using if self-hosting.

Azure DevOps (dev.azure.com)

### If you're self-hosting Renovate, tell us what version of the platform you run.

Azure DevOps

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

We want to update Terraform module versions using the [terraform manager](https://docs.renovatebot.com/modules/manager/terraform/). 

The modules are hosted in Azure DevOps Repositories and are referenced using the [scp ssh pattern](https://www.terraform.io/language/modules/sources#scp-like-address-syntax). 
Clone URLs in Azure DevOps have a syntax like this:
`git@ssh.dev.azure.com:v3/<OrganizationName>/<ProjectName>/<RepositoryName>`

In terraform a reference looks like this:
```
module "some-module" {
  source = "git::git@ssh.dev.azure.com:v3/<OrganizationName>/<ProjectName>/<RepositoryName>//modules/some-module?ref=v0.0.38"
}
```

The current regex pattern in https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/terraform/modules.ts#L18 do not match this clone urls.

The current logs for the module are appended below.

Expected behaviour:
New tags are discovered and can be updated.

### Relevant debug logs

<details><summary>Logs</summary>

```
{
                 "depType": "module",
                 "depName": "git::git@ssh.dev.azure.com:v3/<OrganizationName>/<ProjectName>/<RepositoryName>",
                 "datasource": "terraform-module",
                 "depIndex": 4,
                 "updates": [],
                 "warnings": [],
                 "versioning": "loose",
                 "skipReason": "invalid-value"
               }
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

### PR DESCRIPTION
**Pull request by patst**
## Changes

Allow extraction of urls, organization, project and repository for Azure DevOps SSH Repositories. See #15491 for details

## Context
- closes #15491

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/terraform/modules.spec.ts [hunk 1]
```diff
@@ -115,4 +116,46 @@ describe('modules/manager/terraform/modules', () => {
...
+        'git@ssh.dev.azure.com:v3/MyOrg/MyProject/MyRepository?ref=1.0.0'
+      ).groups;
+      const sshGit = azureDevOpsSshRefMatchRegex.exec(
+        'git::git@ssh.dev.azure.com:v3/MyOrg/MyProject/MyRepository?ref=1.0.0'
+      ).groups;
+      const subfolder = azureDevOpsSshRefMatchRegex.exec(
+        'git::git@ssh.dev.azure.com:v3/MyOrg/MyProject/MyRepository//some-module/path?ref=1.0.0'
+      ).groups;
+
+      expect(ssh.organization).toBe('MyOrg');
```
Comment (viceice) [inline lib/modules/manager/terraform/modules.spec.ts:132]: use `toEqual` object matcher, so we don't need do much function calls

**Comment by patst** [comment]
thanks for the review feedback, updated the code and added another testcase to reach the 100% test coverage

File: lib/modules/manager/terraform/extract.spec.ts [hunk 2]
```diff
@@ -52,6 +53,17 @@ describe('modules/manager/terraform/extract', () => {
...
 
+    it('extracts azureDevOps modules', async () => {
+      const res = await extractPackageFile(
+        azureDevOpsModules,
+        'modules.tf',
+        {}
+      );
+      expect(res.deps).toHaveLength(3);
+      expect(res.deps.filter((dep) => dep.skipReason)).toHaveLength(0);
+      expect(res).toMatchSnapshot();
```
Comment (viceice) [inline lib/modules/manager/terraform/extract.spec.ts:64]: Please use `toEqual` instead, so you only need one `expect`

File: lib/modules/manager/terraform/modules.spec.ts [hunk 3]
```diff
@@ -115,4 +116,61 @@ describe('modules/manager/terraform/modules', () => {
...
+    it('should split organization, project, repository and tag from source', () => {
+      const ssh = azureDevOpsSshRefMatchRegex.exec(
+        'git@ssh.dev.azure.com:v3/MyOrg/MyProject/MyRepository?ref=1.0.0'
+      ).groups;
+      const sshGit = azureDevOpsSshRefMatchRegex.exec(
+        'git::git@ssh.dev.azure.com:v3/MyOrg/MyProject/MyRepository?ref=1.0.0'
+      ).groups;
+      const subfolder = azureDevOpsSshRefMatchRegex.exec(
+        'git::git@ssh.dev.azure.com:v3/MyOrg/MyProject/MyRepository//some-module/path?ref=1.0.0'
+      ).groups;
```
Comment (viceice) [inline lib/modules/manager/terraform/modules.spec.ts:130]: please plit to three tests

File: lib/modules/manager/terraform/modules.spec.ts [hunk 4]
```diff
@@ -115,4 +116,67 @@ describe('modules/manager/terraform/modules', () => {
...
+        url: 'git@ssh.dev.azure.com:v3/MyOrg/MyProject/MyRepository',
+      });
+    });
+
+    it('should parse alpha-numeric characters as well as dots, underscores, and dashes in repo names', () => {
+      const dots = azureDevOpsSshRefMatchRegex.exec(
+        'git::git@ssh.dev.azure.com:v3/MyOrg/MyProject/MyRepository//some-module/path?ref=v1.0.0'
+      ).groups;
+
+      expect(dots.tag).toBe('v1.0.0');
```
Comment (viceice) [inline lib/modules/manager/terraform/modules.spec.ts:171]:
```suggestion
```

File: lib/modules/manager/terraform/modules.spec.ts [hunk 5]
```diff
@@ -115,4 +116,67 @@ describe('modules/manager/terraform/modules', () => {
...
+        repository: 'MyRepository',
+        tag: '1.0.0',
+        url: 'git@ssh.dev.azure.com:v3/MyOrg/MyProject/MyRepository',
+      });
+    });
+
+    it('should parse alpha-numeric characters as well as dots, underscores, and dashes in repo names', () => {
+      const dots = azureDevOpsSshRefMatchRegex.exec(
+        'git::git@ssh.dev.azure.com:v3/MyOrg/MyProject/MyRepository//some-module/path?ref=v1.0.0'
+      ).groups;
```
Comment (viceice) [inline lib/modules/manager/terraform/modules.spec.ts:169]:
```suggestion
      )?.groups;
```

File: lib/modules/manager/terraform/modules.spec.ts [hunk 6]
```diff
@@ -115,4 +116,67 @@ describe('modules/manager/terraform/modules', () => {
       expect(dots.tag).toBe('v1.0.0');
     });
   });
+
+  describe('azureDevOpsSshRefMatchRegex', () => {
+    it('should split organization, project, repository and tag from source url', () => {
+      const ssh = azureDevOpsSshRefMatchRegex.exec(
+        'git@ssh.dev.azure.com:v3/MyOrg/MyProject/MyRepository?ref=1.0.0'
+      ).groups;
```
Comment (viceice) [inline lib/modules/manager/terraform/modules.spec.ts:124]:
```suggestion
      )?.groups;
```

typescript null check

File: lib/modules/manager/terraform/modules.spec.ts [hunk 7]
```diff
@@ -115,4 +116,67 @@ describe('modules/manager/terraform/modules', () => {
...
+        repository: 'MyRepository',
+        tag: '1.0.0',
+        url: 'git@ssh.dev.azure.com:v3/MyOrg/MyProject/MyRepository',
+      });
+    });
+
+    it('should split organization, project, repository and tag from source url with git prefix', () => {
+      const sshGit = azureDevOpsSshRefMatchRegex.exec(
+        'git::git@ssh.dev.azure.com:v3/MyOrg/MyProject/MyRepository?ref=1.0.0'
+      ).groups;
```
Comment (viceice) [inline lib/modules/manager/terraform/modules.spec.ts:139]:
```suggestion
      )?.groups;
```

File: lib/modules/manager/terraform/modules.spec.ts [hunk 8]
```diff
@@ -115,4 +116,67 @@ describe('modules/manager/terraform/modules', () => {
...
+        repository: 'MyRepository',
+        tag: '1.0.0',
+        url: 'git@ssh.dev.azure.com:v3/MyOrg/MyProject/MyRepository',
+      });
+    });
+
+    it('should split organization, project, repository and tag from source url with subfolder', () => {
+      const subfolder = azureDevOpsSshRefMatchRegex.exec(
+        'git::git@ssh.dev.azure.com:v3/MyOrg/MyProject/MyRepository//some-module/path?ref=1.0.0'
+      ).groups;
```
Comment (viceice) [inline lib/modules/manager/terraform/modules.spec.ts:154]:
```suggestion
      )?.groups;
```

**Review by viceice — APPROVED** [review]

**Comment by viceice** [comment]
@patst Please don't forget to add the closes or fixes keyword in pr description to auto close issues in future 😉

**Pull request merged**